### PR TITLE
DoS & Detection Revamp

### PR DESCRIPTION
**Pull request by Swangeon**
Go rewrite of the DoS attack in Honeyscanner.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -99,10 +99,10 @@ The following instructions will guide you through the process of setting up Hone
 Use the following examples as a reference for how to run `Honeyscanner`:
 
 ```bash
-python3 main.py --honeypot cowrie --honeypot_version 2.5.0 --target_ip 127.0.0.1 --port 2222 --username root --password 1234
+sudo python3 main.py --honeypot cowrie --honeypot_version 2.5.0 --target_ip 127.0.0.1 --port 2222 --username root --password 1234
```
Comment (aristofanischionis) [inline README.md:102]: Let's avoid introducing running honeyscanner with `root` privileges, we are just scanning for open ports, so we can change the nmap flags to only '-p-', this way we can keep running honeyscanner as a normal user.

File: .gitignore [hunk 2]
```diff
@@ -31,3 +31,4 @@ Thumbs.db
 *.json
 *.db
 *.txt
+honeyscanner/active_attacks/attacks/attacks
```
Comment (aristofanischionis) [inline .gitignore:34]: too many attacks attacks, nested folders, can we improve this?

File: README.md [hunk 3]
```diff
@@ -99,10 +99,10 @@ The following instructions will guide you through the process of setting up Hone
...
 ```
 ```bash
-python3 main.py --honeypot kippo --honeypot_version 0.9 --target_ip 127.0.0.1 --port 2222
+sudo python3 main.py --honeypot kippo --honeypot_version 0.9 --target_ip 127.0.0.1 --port 2222
```
Comment (aristofanischionis) [inline README.md:105]: same about sudo

File: honeyscanner/active_attacks/attack_orchestrator.py [hunk 4]
```diff
@@ -18,11 +17,11 @@ def __init__(self, honeypot: BaseHoneypot) -> None:
                                      to use in the attacks.
         """
         self.honeypot = honeypot
-        # for dionaea and conpot, we can run the DoSAllOpenPorts attack only
+        # For Dionaea and Conpot, we can run the DoS attack only
```
Comment (aristofanischionis) [inline honeyscanner/active_attacks/attack_orchestrator.py:20]: Is this comment still valid?

File: honeyscanner/active_attacks/honeypot_port_scanner/honeypot_port_scanner.py [hunk 5]
```diff
@@ -48,17 +50,22 @@ def scan_honeypot(self):
...
-                ports[port] = {
+                report_ports[port] = {
                     'name': port_info['service']['name'] if 'name' in port_info['service'] else '',
                     'product': port_info['service']['product'] if 'product' in port_info['service'] else '',
                     'version': port_info['service']['version'] if 'version' in port_info['service'] else '',
                     'cpe': port_info['service']['cpe'] if 'cpe' in port_info['service'] else [],
                     'script': port_info['service']['script'] if 'script' in port_info['service'] else []
                 }
-            self.report['ports'] = ports
+                attack_ports[port] = {
```
Comment (aristofanischionis) [inline honeyscanner/active_attacks/honeypot_port_scanner/honeypot_port_scanner.py:64]: this needs to be adjusted as now we will run a port scan only, not scanning with scripts (-sC) and versions (-sV) which are parts of the (-A) flag. Adjustments need to be made.

File: honeyscanner/main.py [hunk 6]
```diff
@@ -77,8 +85,11 @@ def main() -> None:
     """
     Main entry point of the program.
     """
-    args: argparse.Namespace = parse_arguments()
     print(ascii_art_honeyscanner())
+    if not check_go_exists():
```
Comment (aristofanischionis) [inline honeyscanner/main.py:89]: This is not accurate, maybe go is installed in the system but is not in path, it's more accurate to change the error message to something like go is not found in the path. Also we need to consider go compilation on runtime, I am not sure if it's the best idea. Also, providing the compiled binary is also not the best idea, as it won't work in all architectures. Maybe we can provide for amd64 and arm64 and query the system on runtime and determine which one to use.

**Review by aristofanischionis — CHANGES_REQUESTED** [review]
Hi, thanks for the PR, please review my comments and provide the necessary fixes before merging.

**Comment by aristofanischionis** [comment]
@Swangeon Can you please update README with some FAQs to help users. 
In case of troubles, suggest that they install the latest version of anaconda from here
https://www.anaconda.com/download/success
and that they make sure they use python 3.10
and also if they still face problems with pipenv install, they should try emptying their virtualenvs folder and trying again.

Also, don't we need one more line in Pipfile?
`typing = "*"`
to import the typing library?

**Comment by aristofanischionis** [comment]
DoS attack doesn't seem to be able to take down the honeypot. I am running the Cowrie honeypot in the exact same way I was running it before, with the Python code used before, the honeypot was becoming unavailable, and now with the Go module, not anymore, it is still up.

**Comment by aristofanischionis** [comment]
I received these 2 warnings when trying to run Honeyscanner, did you also experience them? Let's fix them before merging.

Screenshot 2024-07-29 at 14 17 59

**Comment by aristofanischionis** [comment]
The Honeypot port scanner is not working properly for me. 
We can try integrating this PR (https://github.com/honeynet/honeyscanner/pull/38) and removing nmap altogether as it's slow, noisy, and too big for what we need.

Screenshot 2024-07-29 at 14 26 38

**Comment by Swangeon** [comment]
> I received these 2 warnings when trying to run Honeyscanner, did you also experience them? Let's fix them before merging.
> Screenshot 2024-07-29 at 14 17 59

This is going to be an annoying error for now that doesn't effect the project that you can look at more [here](https://github.com/paramiko/paramiko/issues/2419). There are ways to repress the warning, but I don't know if we would want to implement a temporary solution for a temporary problem.

> The Honeypot port scanner is not working properly for me. We can try integrating this PR (#38) and removing nmap altogether as it's slow, noisy, and too big for what we need.
> Screenshot 2024-07-29 at 14 26 38

I haven't gotten this error before so I am not quite sure whats causing it but I will look into it. I'm not quite certain merging this PR with @ZupaPomidorowa's PR would be best for now because this PR would now have to wait for their PR to be merged and functional to work more on this. If you want, I can just move on to the fuzzing upgrades we talked about and come back to this at a later date.

File: README.md [hunk 7]
```diff
@@ -86,7 +86,7 @@ The following instructions will guide you through the process of setting up Hone
 
 - For testing Honeyscanner against [Conpot](https://github.com/mushorg/conpot), use the following Docker Image in DockerHub [here](https://hub.docker.com/r/honeynet/conpot).
 
-- After running a Honeypot using Docker containers locally, you will be able to specify the following parameters: `--target_ip 127.0.0.1 --port 2222` when running the Honeyscanner.
+- After running a Honeypot using Docker containers locally, you will be able to specify the target ip parameter with `--target_ip 127.0.0.1` when running the Honeyscanner.
```
Comment (aristofanischionis) [inline README.md:89]: there is a typo here, correct is to have target-ip not target_ip

File: README.md [hunk 8]
```diff
@@ -99,16 +99,11 @@ The following instructions will guide you through the process of setting up Hone
...
 
 ```bash
-python3 main.py --honeypot cowrie --honeypot_version 2.5.0 --target_ip 127.0.0.1 --port 2222 --username root --password 1234
-```
-```bash
-python3 main.py --honeypot kippo --honeypot_version 0.9 --target_ip 127.0.0.1 --port 2222
-```
-```bash
-sudo python3 main.py --honeypot dionaea --honeypot_version 0.11.0 --target_ip 127.0.0.1 --port 2323
+python3 main.py --target_ip 127.0.0.1
```
Comment (aristofanischionis) [inline README.md:102]: same here

File: README.md [hunk 9]
```diff
@@ -99,16 +99,11 @@ The following instructions will guide you through the process of setting up Hone
...
 ```
+
 ```bash
-sudo python3 main.py --honeypot conpot --honeypot_version 0.6.0 --target_ip 127.0.0.1 --port 2323
+python3 main.py --target_ip 127.0.0.1 --username root --password 1234
```
Comment (aristofanischionis) [inline README.md:106]: same here

File: README.md [hunk 10]
```diff
@@ -127,6 +122,11 @@ Honeyscanner does not have any release yet.
 
 Please use [Github issues](https://github.com/honeynet/honeyscanner/issues) to report any issues or for questions.
 
+## FAQ
```
Comment (aristofanischionis) [inline README.md:125]: i like that!

File: .gitignore [hunk 11]
```diff
@@ -19,6 +19,7 @@ build/
 # IDE files
 .idea/
 .vscode/
+.zed/
```
Comment (aristofanischionis) [inline .gitignore:22]: what is this for?

File: .gitignore [hunk 12]
```diff
@@ -28,6 +29,7 @@ build/
 Thumbs.db
 
 # misc
-*.json
 *.db
+*.db-journal
```
Comment (aristofanischionis) [inline .gitignore:33]: what is this for?

**Review by aristofanischionis — CHANGES_REQUESTED** [review]
I am facing some issues while running this version of Honeyscanner
Screenshot 2024-08-22 at 14 53 35
Screenshot 2024-08-22 at 14 53 46
Screenshot 2024-08-22 at 14 53 59

and dos attack seems to not be working anymore

**Review by aristofanischionis — APPROVED** [review]
Thank you for the great work and completing successfully Google Summer of Code 2024. We are glad we had you as part of our team!
Good luck in your future endeavours.

**Pull request merged**